### DISCUSSION TITLE
Documentation/examples [ongoing discussion thread]

### DISCUSSION THREAD

**Discussion by simendsjo** [leppie/IronScheme#86] [Discussions]
I'm having a hard time finding any documentation on how to use this. Apart from the examples in the repository, I'm turing up blank. And as this is using a fork of dlr, documentation for that project doesn't seem to be valid either.

Are there any documentation floating around that I'm unable to locate?

**Comment by leppie** [comment]
First stop would be the R6RS docs. For some libraries, like SRFI's, they tend to be already well documented online. Manuals like TSPL3 covers R6RS in depth too, but it does have some Chez Scheme specific stuff.

The DLR is not exposed and not meant to be used externally with IronScheme. Dont worry about that.

I used to have some example projects, but I killed them when I moved the repo a little while back. (mental note: add them as seperate repos).

For specific examples on calling .NET code from Scheme, grep the code for `clr-` to find. (IIRC this was on the old codeplex site, is that even still around? need to look).

I know this is not the answer you are looking for, but generally better to ask a specfic question :D

**Comment by leppie** [comment]
Would seem the wiki on codeplex has been hosed :(

**Comment by leppie** [comment]
I still have a doc listing site up at http://ironscheme.net/doc

Basically will show proc forms, macros, etc

It is quite outdated though, but not much has changed.

**Comment by simendsjo** [comment]
Thanks! I was looking at the possibility of using IronScheme as an embedded scripting language for a dotnet project. In such case, I would like to heavily restrict what libraries and functions are available (no filesystem, no IO etc). And I would like a way to extend what is available.

Perhaps IronScheme is a bad match to use for an embedded scripting language? I'm also looking at schemy which might be a better choice, but on the other hand is *very* bare-bones and would require quite some work.

**Comment by leppie** [comment]
I have given some thought to your requirements.

Basically the interface from C# to IronScheme is just an `Eval` method taking a 'code' string and the important part, an environment. An environment is basically just a bunch of library imports, and the code being eval'd will only use that environment.

Normally the environment is not specified, and it uses what is called the interaction-environment which is mutable (basically the same a REPL would have). By default this includes the `(ironscheme)` library which is a superset of `(r6rs)`. 

So if you can craft a custom interaction-environemnt (in your case, some library exposing just what you want to), should be sufficient to lock down access to ports, and specifically importing other libraries that are not 'authorized'.

I have not played with this idea, not sure if it is possible in its current state, but it should be trivial to add something.

**Comment by bcalco** [comment]
I also plan to use IronScheme (alongside IronPython) to give users of my application extensibility options suitable for a range of skill levels (newbie-devs=> python, experienced-devs => scheme). My intent is not so restrictive as simendsjo's. There are scenarios where I plan for users to provide well-defined functions in specific contexts with environments I plan to pass into these functions, and others where I encourage them to script freely user-defined tasks and procedures of arbitrary complexity.

I was a bit confused at first as well how to begin, but my experience with IronScheme so far (such as working through examples in classic Scheme texts, and studying the examples and libraries that ship in the release) is encouraging, especially compared to my first choice, ClojureCLR, which is even more undocumented and unclear how to integrate in real .NET applications. I'm slowly figuring that out, too, but Scheme will likely be a perfectly sufficient Lisp for my purposes, until/unless ClojureCLR gets a lot more stable and usable.

My main concern is performance on the DLR, and just whether/how far support for it continues to this day. I am also concerned about/interested in running on mono/dotnet core on Mac/Linux, so designing my application (or at least parts of it) to run cross platform even if other parts must run on Windows is something I'm noodling on. However, if necessary, I'm prepared to get involved in maintaining the DLR, and whichever DLR-based scripting engines I decide to embed.

I think some tutorials on the github wiki would be helpful. I'm not sure if adding them to my fork and issuing pull requests will work for wiki entries as it does for code (I'm assuming the answer is yes, but never tried it actually). If so, I'm keen to work on some basic wiki pages to help others get through the learning curve as I surf it myself.

**Comment by leppie** [comment]
@bcalco As for performance, not much is based on the DLR anymore. I specifically tweaked codegen for for IronScheme. To be honest, the DLR I based it on is very different from what it is now. All code is compiled, nothing interpreted. 

IronScheme runs pretty much any .NET. .NET desktop 3.5/4.0+, Mono, .NET Core 2.1+, tested on Windows and Linux all in a single DLL :)

**Comment by bcalco** [comment]
@leppie I am pretty jazzed that I found it. The unstable, seemingly unmaintained state of ClojureCLR is what sent me looking. I am planning to build something quite special with it and am glad to hear the DLR's not so central to it from a performance drag point of view. I would however like to know how to compile DLL and EXE artifacts written in IronScheme, if that's possible (like it is for both F# and IronPython).

**Comment by leppie** [comment]
@bcalco compiling libraries is pretty simple, this is however only supported on .NET desktop (and possibly Mono). Any version/bitness should be ok. They should be loadable in any version/bitness too.

The steps are:
1. Create a library say test-lib.sls with the following:
```
(library 
  (test-lib) 
  (export foo) 
  (import (rnrs)) 

(define (foo) 'bar))
```
2. In the REPL, import the library ie `(import (test-lib))`, the call `(compile)`

You should get something like:
![image](https://user-images.githubusercontent.com/127666/76993969-60c74f00-6956-11ea-8d07-ac3094ef51a3.png)

Now whenever you import the library, it will load the compiled version first if it is not out of date.

When out of date or a dependent library has been recompiled you will get a warning like:
![image](https://user-images.githubusercontent.com/127666/76994247-c0bdf580-6956-11ea-8fee-9dc2c08f7430.png)

As you can see, the library is now still loaded.

Also when calling `(compile)` it will compile any imported currently in the REPL. It will as such also compile any dependent library if needed.

Side note: You cant call compile again in the same REPL session when DLL's have been loaded. If that ever happens, just delete the DLL's and call compile again. IronScheme is fast enough that compiling should really only be necesary when creating a final version. Also these compiled libraries will only load on the exact version of IronScheme that compiled it.

**Comment by leppie** [comment]
Just a note for those looking for docs, specifically for Scheme. The R6RS spec is very good and contains nice examples. For more in depth stuff you can look at TSPL4. 

IronScheme also shares a lot in common with the following Schemes:
* Ikarus
* Vicare
* Chez Scheme

**Comment by bcalco** [comment]
@leppie What about compiling EXE targets, with a canonical 'main' function accepting command line arguments?

**Comment by leppie** [comment]
@bcalco There is no support to compile a program as such, but nothing stops you from making a `main` procedure in a library and then simply say invoke it from C#. Example:

```
using IronScheme;
...
"(import (my-app))".Eval();
"(main {0})".Eval(Cons.FromList(argsArray));
```

Here `main` might be defined as:
```
(define (main args)
  ;; your stuff
)
```

**Comment by bcalco** [comment]
@leppie I'd be very interested to see (and help add) support for generating EXEs and DLLs from IronScheme scripts and libraries, respectively. IronPython's `ipyc.exe` is a good model:

```
PS C:\IronPython 2.7> ipyc
ERROR: No files or main defined
ipyc: The Command-Line IronPython Compiler

Usage: ipyc.exe [options] file [file ...]

Options:
    /out:output_file                          Output file name (default is main_file.<extenstion>)
    /target:dll                               Compile only into dll.  Default
    /target:exe                               Generate CONSOLE executable stub for startup in addition to dll.
    /target:winexe                            Generate WINDOWS executable stub for startup in addition to dll.
    /py:<option>                              Allows specifying options that modify the behavior of IronPython (e.g., -X:FullFrames)
    @<file>                                   Specifies a response file to be parsed for input files and command line options (one per line)
    /? /h                                     This message

EXE/WinEXE specific options:
    /main:main_file.py                        Main file of the project (module to be executed first)
    /platform:x86                             Compile for x86 only
    /platform:x64                             Compile for x64 only
    /embed                                    Embeds the generated DLL as a resource into the executable which is loaded at runtime
    /standalone                               Embeds the IronPython assemblies into the stub executable.
    /mta                                      Set MTAThreadAttribute on Main instead of STAThreadAttribute, only valid
for /target:winexe
    /errfmt:msg                               A string that will be used when showing an error occured, {{0}} will be replaced by the exception message
    /win32icon:file.ico                       Sets file.ico as the icon for the executable

Example:
    ipyc.exe /main:Program.py Form.py /target:winexe
PS C:\IronPython 2.7>
```

That said, thanks for the suggested workaround. In the case where C# app is calling a specific IronScheme script in this way, I'd be interested to understand what additionally can be done on the C# side to expose C# application abstractions to the script (and any scripts in imports/loads), such that it can access it as built-in. Is the host app's namespace(s) already imported into or otherwise available to the script's run time "image" so that it can define wrapper functions around it via (import ...)? 

Also, let me just say, as I've been exploring all the scripts and libraries that come with IronScheme, I really appreciate the incredible work that's gone into making it a full, proper Scheme. I would love to see IronScheme attract more developer mind share in both the Scheme and the .NET worlds.

**Comment by leppie** [comment]
> Also, let me just say, as I've been exploring all the scripts and libraries that come with IronScheme, I really appreciate the incredible work that's gone into making it a full, proper Scheme. I would love to see IronScheme attract more developer mind share in both the Scheme and the .NET worlds.

Thanks, I appreciate it

(will reply to questions tomorrow)

**Comment by leppie** [comment]
But a quick answer: Everything in IronScheme happens via scheme code.

* compiling (procedure)
* referencing assemblies (syntax)
* importing namespaces (syntax)
* enabling debug code generation (procedure) 

Also R6RS's module system takes care of all (direct or indirect) library dependencies.

PS: If you enable debug code generation you can literally debug scheme scripts right in VS (it is not great, but not bad either).

**Comment by bcalco** [comment]
I started my Lisp journey with Clojure, which I still quite enjoy apart from the business of running on the JVM. The main draw was its philosophy of reducing incidental complexity (running on the JVM notwithstanding). From there I dabbled in Common Lisp but was disappointed not so much by the language (which has really captured my imagination) as by the uneven quality of implementations. Among the open source implementations, SBCL stands out, but it still lacks full support on Windows, and despite QuickLisp making library distribution easy and Roswell making it simple to try different flavors, there is no UI integration that is anything but a half-done weekend project. Commercial CLs are even more disappointing. Not only are their feature sets stuck in the 1990's, their pricing models are stuck in the 1980's. LispWorks charges $5k per dev per platform upfront (no royalties). Franz negotiates your initial price (like a drug dealer), but eventually gets you to the $7K per dev per platform price, plus 5-7% royalties. Egads, who does that today? The UI's suck, and to get basic database access (which on LispWorks amounts to using ODBC, bah) you also have to accept CORBA as a major enterprise feature (I guess they're still trying to get their ROI on that). WTH?!

Oracle changing licensing for commercial use of the JVM was a game changer for me. It means if I write commercial software for the desktop using Clojure (or ABCL), desktop users now have to pay Larry Ellison ~$20/mo for the privilege of running my software. (I can only imagine what it means for cloud providers, but they're probably getting around it by using OpenJDK.) Meanwhile, .NET finally got un-stuck off Windows-only implementations, and meanwhile open source projects have gone the direction of choosing more commercial friendly licenses (MIT, BSD, Apache, etc). And UIs are nice, with lots of choices (winforms, wpf, avalonia, eto, gtksharp, etc). 

If I had immense free time and was sitting on a pile of cash, I might consider helping SBCL integrate consistently across platforms with, say, wxWidgets and create an IDE to compete with the commercial players. But for now, I have to build software people will buy, and I'm quite content to use a Lisp on .NET. I thought it would be ClojureCLR, but it looks like IronScheme is more ready and stable for my immediate needs. I really appreciate your responsiveness, too. I think this can work.

**Comment by leppie** [comment]
@bcalco Anything helps :D 

I understand the lack of docs specific to IronScheme (basically not covered by R6RS and .NET) is a hurdle for many wanting to start out. 

There used to stuff on Codeplex, but that got shutdown. There is doc stuff in the repo, which is horribly outdated. There is the problem that I cant even touch type after 20 years of coding :D 

I have probably repeated most of the functionality at some point, but over the last 13 years it seems to have gotten lost or hidden.

**Comment by bcalco** [comment]
How might one inspect a doc string comment in IronScheme? Suppose the following function:

```scheme
> (define (factorial n)
.   "Computes the factorial of a given integer value."
.   (if (= 0 n) 1
.       (* n (factorial (- n 1)))))
> (factorial 5)
120
> (factorial 1000)
402387260077093773543702433923003985719374864210714632543799910429938512398629020592044208486969404800479988610197196058631666872994808558901323829669944590997424504087073759918823627727188732519779505950995276120874975462497043601418278094646496291056393887437886487337119181045825783647849977012476632889835955735432513185323958463075557409114262417474349347553428646576611667797396668820291207379143853719588249808126867838374559731746136085379534524221586593201928090878297308431392844403281231558611036976801357304216168747609675871348312025478589320767169132448426236131412508780208000261683151027341827977704784635868170164365024153691398281264810213092761244896359928705114964975419909342221566832572080821333186116811553615836546984046708975602900950537616475847728421889679646244945160765353408198901385442487984959953319101723355556602139450399736280750137837615307127761926849034352625200015888535147331611702103968175921510907788019393178114194545257223865541461062892187960223838971476088506276862967146674697562911234082439208160153780889893964518263243671616762179168909779911903754031274622289988005195444414282012187361745992642956581746628302955570299024324153181617210465832036786906117260158783520751516284225540265170483304226143974286933061690897968482590125458327168226458066526769958652682272807075781391858178889652208164348344825993266043367660176999612831860788386150279465955131156552036093988180612138558600301435694527224206344631797460594682573103790084024432438465657245014402821885252470935190620929023136493273497565513958720559654228749774011413346962715422845862377387538230483865688976461927383814900140767310446640259899490222221765904339901886018566526485061799702356193897017860040811889729918311021171229845901641921068884387121855646124960798722908519296819372388642614839657382291123125024186649353143970137428531926649875337218940694281434118520158014123344828015051399694290153483077644569099073152433278288269864602789864321139083506217095002597389863554277196742822248757586765752344220207573630569498825087968928162753848863396909959826280956121450994871701244516461260379029309120889086942028510640182154399457156805941872748998094254742173582401063677404595741785160829230135358081840096996372524230560855903700624271243416909004153690105933983835777939410970027753472000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
> factorial
#<procedure factorial>
```

I'm interested to inspect the function and ideally have it present the doc string comment to me that I supplied in its definition. I've tried variations of `(documentation ...)` and `(inspect ...)` but nothing seems to work. The value returned by simply inputting the function name is `#<procedure factorial>` without its documentation string.

**Comment by leppie** [comment]
@bcalco You have been spoilt by Clojure :D There is nothing built in.

You can achieve this with macro's though. Let me try dish up something simple.

**Comment by leppie** [comment]
Here is a terribly inefficient, but working version (just paste in REPL)

```
(import 
  (ironscheme)
  (ironscheme syntax utils))

(define-syntax doc
  (lambda (x)
    (syntax-case x ()
      [(_ proc)
        (lambda (lookup)
          (lookup (syntax-format "doc:~a" #'proc #'proc)))])))  

(define-syntax def
  (lambda (x)
    (syntax-case x ()
      [(_ (name arg ...) doc-str b b* ... )
        (string? (syntax->datum #'doc-str))
        (with-syntax [(doc-id (syntax-format "doc:~a" #'name #'name))]
          #'(begin
              (define-syntax doc-id (make-compile-time-value 'doc-str))
              (define (name arg ...) b b* ...)))]
      [(_ (name arg ...) b b* ... )
        #'(define (name arg ...) b b* ...)])))

(def (fac n) "hello world" n)

(fac 2) ; prints 2

(doc fac) ; prints "hello world"
```

**Comment by bcalco** [comment]
@leppie Thanks - I'm still getting my head around Scheme macros (vs CL vs Clojure), this is actually quite helpful.  Why didn't you call the new function macro `defn` ...  Ahem. Just kidding! ;-)

Honestly I was mildly surprised `(define ...)` let me inject a document comment in the definition.  I haven't really seen that used in Scheme code, certainly not like it is in CL and Clojure anyway. So I thought, if it's there "Perhaps there are some functions or macros for inspecting such doc strings in IronScheme..."

**Comment by leppie** [comment]
@bcalco That macro I wrote is pretty advanced stuff, but it is nice simplistic sample showing a lot of features :D

The doc string in that position is just loaded and discarded, so that is why it works.

Also, nice example showing hygiene in action:
```
> (def (f) "f" (def (f) "inner f" 1) (doc f))
> (doc f)
"f"
> (f)
"inner f"
> (doc f)
"f"
```

You can also get some info about the procedure by introspection:
```
> (def (k a b c) d)
> (procedure-form k)
(k a b c)
```

Note: `procedure-form` is not a macro, it expects a procedure arguement, while the `doc` macro binds to an identifier. You can experiment if you want to maybe drop the doc string into runtime, but that is up to you.

**Comment by bcalco** [comment]
Implementing a full Scheme macro system compiling to IL must have been quite an experience. Did you ever wish Scheme was a LISP-2 while working on it? (AFAIK, Clojure is also a LISP-1, but I'm unaware that its macro system deals with "hygiene" besides supporting `(gensym)`... ). I'm curious how you look at CL macros vs Scheme macros in light of having implemented Scheme's system.

**Comment by leppie** [comment]
@bcalco I have never worked with CL. The macro system is written in Scheme and it was not written by me. But it is just Scheme that is run :D

**Comment by leppie** [comment]
Good news! It looks like the Wayback Machine has a copy of the old codeplex  docs !

https://web.archive.org/web/20160421103033/http://ironscheme.codeplex.com/documentation

**Comment by leppie** [comment]
@bcalco FYI, this is why doc-string are not in Scheme yet. It is a lot more complicated than in Clojure, and many people have to concur instead of just 1 🗡️ 

https://groups.google.com/forum/#!topic/scheme-reports-wg2/o_Cnqjbdkqo

**Comment by bcalco** [comment]
@leppie I actually joined that google group a week or so ago and noted with interest the discussion on doctrings. Design by committee is not fun! But it is still useful to hear all the perspectives and the issues they raise. I am, at this point, very committed to working in and with IronScheme. I have a product I want to build with it, and I will have some free time (I hope) this weekend to play with it. I still really want to be able to generate EXEs, but I'll start off with a "host" app in C# that calls a main script, which in turn loads what-else it needs. I would like to understand about how a `(define (main args) ...)`-based "app" loaded by a host C# EXE could make host-level abstractions (including possibly forms and utility classes) available to the script part of the app---was this clear, or do I need to elaborate? (I feel like I need to elaborate, so I'll give this request some thought and add another comment when I think I can explain it more concretely.) Also, should I enter an enhancement request for compiling EXEs, or is this something you'll only consider at the point of a gun (or the receiving end of a very big bank wire)? (I have only the former, but alas, you're on another continent, so I have to rely on your kind nature.) :)

**Comment by bcalco** [comment]
Also... Assuming I want to embed IronScheme in an application, what assumptions should I make about packaging it with my application (and any libraries whose source I share alongside IronScheme? And is it possible to precompile all the libraries as DLLs and include them as DLLs in my project? IOW, what are the configuration options.

**Comment by leppie** [comment]
> I would like to understand about how a `(define (main args) ...)`-based "app" loaded by a host C# EXE could make host-level abstractions (including possibly forms and utility classes) available to the script part of the app---was this clear, or do I need to elaborate? (I feel like I need to elaborate, so I'll give this request some thought and add another comment when I think I can explain it more concretely.)

It is pretty clear, but still it is very broad. I think a sample app or a few might explain it better.

**Comment by bcalco** [comment]
@Leppie If a "main" script such as hypothesized above is started by a C# application, does it automatically have access to all the namespaces in the host application (and any loaded DLLs), or does the host app have to invoke some API calls to make specific namespaces/classes etc. available to the script (and its dependencies)?

And by "automatically" I mean requiring only `(import ...)` statements, or does it need to use clr functions to load them independently?

**Comment by bcalco** [comment]
Let's imagine one wants to write Schemacs, Emacs using IronScheme instead of Elisp. The controlling C# app has a bunch of stuff (e.g., some concept of a Buffer) that the Schemacs Lisp engine will need to refer to. Some of these will be built into the EXE, some may be provided by dependent DLLs that the host app loads as part of its startup. In the ideal case scenario, the `(main...) ` script (and any script libraries it imports) just need to `(import ..)` those namespaces, instead of use CLR functions to load them explicitly (again).
